### PR TITLE
Build experiments

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @mmitche \ @natemcmaster

**Comment by natemcmaster** [comment]
I think you also need to update https://github.com/dotnet/dotnet-ci/blob/master/data/repolist.txt to get this repo to build on dotnet-ci.

**Comment by pranavkm** [comment]
Yup, I need to get @Eilon to set up the webhooks too. Just getting this out there to see if this is authored correctly.

File: pipeline.groovy [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import org.dotnet.ci.pipelines.Pipeline
+
+def windowsPipeline = Pipeline.createPipeline(this, 'build/buildpipeline/windows.groovy')
+def linuxPipeline = Pipeline.createPipeline(this, 'build/buildpipeline/linux.groovy')
+def osxPipeline = Pipeline.createPipeline(this, 'build/buildpipeline/osx.groovy')
+String configuration = 'Release'
+
+windowsPipeline.triggerPipelineOnPush(['Configuration':configuration])
```
Comment (mmitche) [inline pipeline.groovy]: Did you not want this for PRs?

**Review by mmitche — APPROVED** [review]

**Comment by natemcmaster** [comment]
Do we need 4 .groovy files? Or can they be merged?

File: pipeline.groovy [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import org.dotnet.ci.pipelines.Pipeline
+
+def windowsPipeline = Pipeline.createPipeline(this, 'build/buildpipeline/windows.groovy')
+def linuxPipeline = Pipeline.createPipeline(this, 'build/buildpipeline/linux.groovy')
+def osxPipeline = Pipeline.createPipeline(this, 'build/buildpipeline/osx.groovy')
```
Comment (natemcmaster) [inline pipeline.groovy]: nit: macos ?

**Comment by mmitche** [comment]
I wouldn't merge them too much.  Parameterizing also tends to lead to hard to read scripts too (and really long job names, since the job name is generated off of the input parameters)

File: pipeline.groovy (hunk 1 above)
Comment (pranavkm) [reply pipeline.groovy]: Yes. Do I need to call `triggerPipelineOnEveryGithubPR`?

File: pipeline.groovy (hunk 2 above)
Comment (pranavkm) [reply pipeline.groovy]: Sure.

File: pipeline.groovy (hunk 1 above)
Comment (mmitche) [reply pipeline.groovy]: Yes

**Comment by mmitche** [comment]
You'll also need to update the repolist, add webhooks for the ci3 server, and ensure dotnet-bot has write access to the repo to update status.  I can those things if you give me admin access.

**Comment by pranavkm** [comment]
> add webhooks for the ci3 server, 

Waiting for @Eilon to become accessible to do this. 

> and ensure dotnet-bot has write access to the repo to update status.

What does it update?

**Comment by mmitche** [comment]
The commit status (which shows up in a PR).  IIRC you have to have write access to do this.  The PAT is only a repo:status PAT.  I can try without r/w first.

File: build/buildpipeline/pipeline.groovy [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import org.dotnet.ci.pipelines.Pipeline
```
Comment (pranavkm) [inline build/buildpipeline/pipeline.groovy]: Moved this file to inner dir and added `triggerPipelineOnEveryGithubPR`

**Review by natemcmaster — APPROVED** [review]
LGTM

**Comment by mmitche** [comment]
@dotnet-bot test ci please

**Comment by pranavkm** [comment]
@dotnet-bot test ci please

**Comment by pranavkm** [comment]
Doesn't look like the bot's looking at this What are we missing?

**Comment by mmitche** [comment]
@pranavkm still missing the webhooks I think.

**Comment by Eilon** [comment]
@mmitche sent me some more info so I'll dig into it.

**Comment by Eilon** [comment]
@dotnet-bot test ci please

**Comment by Eilon** [comment]
I saw the WebHook trigger but I don't see the bot doing anything... any ideas on how to debug this?

**Comment by mmitche** [comment]
@Eilon dotnet-bot needs write access to the repo in order to update repo commit status.

**Comment by Eilon** [comment]
@dotnet-bot test ci please

**Comment by mmitche** [comment]
@Eilon still no access.

**Comment by Eilon** [comment]
I gave it push access.. I'll try admin...

**Comment by Eilon** [comment]
@dotnet-bot test ci please

File: build/buildpipeline/pipeline.groovy [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+def windowsPipeline = Pipeline.createPipeline(this, 'windows.groovy')
```
Comment (natemcmaster) [inline build/buildpipeline/pipeline.groovy]: Ok, learned some things from getting aspnet-docker working. it looks like the second parameter needs to be the file path relative to repo root., so `Pipeline.createPipeline(this, 'build/buildpipeline/windows.groovy')`

File: build/buildpipeline/pipeline.groovy [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+def linuxPipeline = Pipeline.createPipeline(this, 'linux.groovy')
+def osxPipeline = Pipeline.createPipeline(this, 'osx.groovy')
+String configuration = 'Release'
+
+[windowsPipeline, linuxPipeline, osxPipeline].each { pipeline ->
+    pipeline.triggerPipelineOnEveryGithubPR(['Configuration':configuration])
```
Comment (natemcmaster) [inline build/buildpipeline/pipeline.groovy]: Missing an argument. This method requires you specify a trigger name. Example:
```groovy
pipeline.triggerPipelineOnEveryGithubPR("Windows ${configuration} x64 Build", ['Configuration':configuration])
```

Comment (pranavkm) [reply build/buildpipeline/pipeline.groovy]: Ok. Do you need the empty groovy files to be present before it can build?

Comment (natemcmaster) [reply build/buildpipeline/pipeline.groovy]: I think so. I added sub files to the dev branch of aspnet-docker first to get this rolling. That triggers the "generator" build. Example https://ci3.dot.net/job/aspnet_aspnet-docker/job/dev/job/generator/

Comment (mmitche) [reply build/buildpipeline/pipeline.groovy]: The generator reads the job declarations upon each commit.  The job definition (code that actually runs) is executed based on the code in the PR/branch (for commit jobs)

Comment (pranavkm) [reply build/buildpipeline/pipeline.groovy]: Ok. Pushed an update

**Comment by pranavkm** [comment]
@dotnet-bot test ci please

**Comment by pranavkm** [comment]
Guess it's running in to permission issues over here too - https://ci3.dot.net/job/aspnet_aspnet-docker/job/dev/job/linux-pipeline-folderFilter_2.0___prtest/lastFailedBuild/console?

**Comment by mmitche** [comment]
Yes.  I'm still unsure why dotnet-bot is unable to update the status for those repos

**Comment by pranavkm** [comment]
@dotnet-bot test ci please

**Comment by pranavkm** [comment]
Yaay!

**Comment by Eilon** [comment]
Woohoo! Thanks @mmitche !!!

**Comment by pranavkm** [comment]
Although it doesn't look like it actually built anything. Is that just a scripting error from our end?

**Comment by natemcmaster** [comment]
IIUC that's because you need commit build/buildpipeline/pipeline.groovy first to the dev branch. Job _definitions_ can be modified via PR, but not job _declarations_.

**Comment by pranavkm** [comment]
Gotcha. Let me get this PR commited in that case. We can iterate on it.

**Pull request merged**